fendermint/vm/topdown/examples/run_polling.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+use async_stm::{atomically_or_err, StmDynError};
+use fendermint_vm_topdown::{
+    Config, Error, IPCAgentProxy, IPCParentFinality, InMemoryFinalityProvider,
+    ParentFinalityProvider, ParentViewProvider, PollingParentSyncer,
+};
+use fvm_shared::address::{set_current_network, Network};
+use ipc_agent_sdk::apis::IpcAgentClient;
+use ipc_agent_sdk::jsonrpc::JsonRpcClientImpl;
+use ipc_sdk::subnet_id::SubnetID;
+use num_traits::FromPrimitive;
+use std::str::FromStr;
+use std::sync::Arc;
+use std::time::Duration;
+
+#[tokio::main]
+async fn main() {
+    set_network_from_env();
+
+    tracing_subscriber::fmt()
+        .with_max_level(tracing::Level::INFO)
+        .with_target(false)
+        .init();
+
+    let url = std::env::var("URL").unwrap_or_else(|_| "http://0.0.0.0:3030/json_rpc".to_string());
+    let raw_target_subnet_id = std::env::var("TARGET").unwrap_or_else(|_| "/r31415926".to_string());
+    let subnet = SubnetID::from_str(&raw_target_subnet_id).unwrap();
+
+    let json_rpc = JsonRpcClientImpl::new(url.parse().unwrap(), None);
+    let ipc_agent_client = IpcAgentClient::new(json_rpc);
+    let agent_proxy = IPCAgentProxy::new(ipc_agent_client, subnet).unwrap();
+
+    let chain_head = agent_proxy.get_chain_head_height().await.unwrap();
+
+    let config = Config {
+        chain_head_delay: 10,
+        polling_interval_secs: 5,
+    };
+    let provider = InMemoryFinalityProvider::new(
+        config.clone(),
+        IPCParentFinality {
+            height: chain_head - 20,
+            block_hash: vec![],
+        },
+    );
+    let provider = Arc::new(provider);
+    let agent = Arc::new(agent_proxy);
+    let polling = PollingParentSyncer::new(config, provider.clone(), agent);
+
+    tokio::spawn(async move {
+        polling.start().unwrap();
+    });
+
+    loop {
+        atomically_or_err(|| {
+            let proposal = match provider.next_proposal() {
+                Ok(p) => p,
+                Err(e) => match e {
+                    StmDynError::Abort(e) => {
+                        let error = e.downcast_ref::<Error>().unwrap();
+                        match error {
+                            Error::HeightNotReady => {
+                                println!("polling not started yet");
+                                return Ok(());
+                            }
+                            _ => unreachable!(),
+                        }
+                    }
+                    _ => unreachable!(),
+                },
+            };
+            println!("proposal: {proposal:?}");
+
+            if let Some(p) = proposal {
+                let msgs = provider.top_down_msgs(p.height)?;
+                println!("topdown messages: {:?}", msgs);
+            }
+
+            Ok(())
+        })
+        .await
```

Comment:
with the async-stm 0.4, there is no need to do the matching anymore, updated already.

Reply:
Cool. By the way if within STM you want to do A, get out of it for some reason and do B instead, then the combinators you want to use are `retry` and `or.

Here's an [example](https://github.com/aakoshh/async-stm-rs/blob/v0.4.0/src/test.rs#L209-L216), which is a copy of a test scenario else implemented in a different system, and I wanted to show that it works here. It's a bit confusing because I use `abort` with an error normally, and here it's just an alias for `retry`. But there you go, with retry it backs out, `or` discards the changes and goes into the other branch.